File: DFS&BFS/programmers/[Lv2] 게임 맵 최단거리/main.py
def solution(maps):
    m = len(maps[0])
    n = len(maps)
    
    visit = [[0 for _ in range(m)] for _ in range(n)] # 방문 여부 
    dx = [0, 0, -1, 1]
    dy = [-1, 1, 0, 0]
    
    queue = [(0, 0)] # (x, y)
    visit[0][0] = 1 # 방문 표시
    
    while queue:
        curX, curY = queue.pop(0)
        for idx in range(4):
            x = curX + dx[idx]
            y = curY + dy[idx]
            
            if x < 0 or x >= n or y < 0 or y >= m:
                continue
            if maps[x][y] == 0: # 벽이면
                continue
            if visit[x][y] != 0: # 방문했으면
                continue
            
            queue.append((x, y))
            visit[x][y] = visit[curX][curY] + 1
    
    return visit[n-1][m-1] if visit[n-1][m-1] != 0 else -1

File: DFS&BFS/programmers/[Lv2] 게임 맵 최단거리/main.cpp
/*
    - Input
        - maps = 게임 맴의 상태 
            - n x m (1 이상 100 이하의 자연수)
            - 0과 1로만 이루어짐, 0은 벽이 있는 자리 1은 벽이 없는 자리
            - 캐릭터는 (1,1)에 상대팀은 (n,m)에 위치 
    - Output
        - 상대 팀 진영에 도착하기 위해서 지나가야 하는 칸의 개수의 최솟값
    - Hint
        - BFS는 1칸씩 주변을 탐색함. 
        - 즉 같은 레벨의 노드를 탐색하기 때문에 (트리라면 같은 높이, 그래프 혹은 행렬이라면 같은 거리) 최단거리를 탐색할 수 있다
        - BFS는 큐를 사용한다는 것에 집중하자!!!
*/
#include<vector>
#include<queue>
#include<algorithm>

using namespace std;

int dx[4] = {-1, 0, 1, 0};
int dy[4] = {0, -1, 0, 1};

int solution(vector<vector<int> > maps)
{
    int n = maps.size();
    int m = maps[0].size();
    int answer = 0;
    queue<pair<int,int>> q;
    vector<vector<int>> vis(n+1, vector<int>(m+1, -1));
    
    q.push({1, 1});
    vis[1][1] = 1;
    
    while(!q.empty()){
        auto curr = q.front();
        q.pop();
        
        for(int i=0; i <4; i++){
            int x = curr.first + dx[i];
            int y = curr.second + dy[i];
            
            if (x <= 0 || x > n || y <= 0 || y > m) continue;
            if (vis[x][y] != -1 || maps[x-1][y-1] == 0) continue;
            
            q.push({x, y});
            vis[x][y] = vis[curr.first][curr.second] + 1;
        }
        
    }
    
    return vis[n][m];
}